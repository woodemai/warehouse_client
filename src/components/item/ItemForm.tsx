import {
    Dialog,
    DialogContent,
    DialogDescription,
    DialogHeader,
    DialogTitle,
} from "@/components/ui/dialog"
import { useForm } from "react-hook-form"
import { zodResolver } from "@hookform/resolvers/zod"
import * as z from "zod"
import axios from "axios"
import {
    Form,
    FormControl,
    FormDescription,
    FormField,
    FormItem,
    FormLabel,
    FormMessage,
} from "@/components/ui/form"
import { Input } from "@/components/ui/input"
import { Button } from "@/components/ui/button"
import { Calendar } from "@/components/ui/calendar"
import {
    Popover,
    PopoverContent,
    PopoverTrigger,
} from "@/components/ui/popover"
import { format } from "date-fns"
import { Calendar as CalendarIcon, Check, ChevronsUpDown } from "lucide-react"

import { cn } from "@/lib/utils"
import { ScrollArea } from "@/components/ui/scroll-area"
import { FC, useState } from "react"
import { CategoryProps } from "../category/Category"
import {
    Command,
    CommandEmpty,
    CommandGroup,
    CommandInput,
    CommandItem,
} from "@/components/ui/command"
import formSchema from "./ItemFormSchema";
import { SupplierProps } from "../supplier/Supplier"


interface ItemFormProps {
    categories: CategoryProps[],
    suppliers: SupplierProps[],
    updating?: boolean,
    id?: string
    name?: string,
    description?: string,
    productionDate?: string,
    expirationDate?: string,
    storageCondition?: string
    weight?: number
    price?: number,
    category?: CategoryProps,
    supplier?: SupplierProps
}
const ItemForm: FC<ItemFormProps> = ({
    categories,
    suppliers,
    updating,
    id,
    name,
    description,
    supplier,
    storageCondition,
    weight,
    price,
    category
}) => {
    const [isOpen, setIsOpen] = useState(false);
    const form = useForm<z.infer<typeof formSchema>>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            name: name ?? "",
            description: description ?? "",
            supplier: supplier?.id,
            storageCondition: storageCondition ?? "",
            productionDate: new Date(),
            expirationDate: new Date(),
            category: category?.id,
            weight: weight ?? 0,
            price: price ?? 0
        },
    })
    const onSubmit = (values: z.infer<typeof formSchema>) => {
        updating
            ? axios.put(`http://localhost:8080/item/${id}`, { ...values })
            : axios.post("http://localhost:8080/item", { ...values })
        form.reset()
        setIsOpen(false)

    }
    return (
        <Dialog open={isOpen} onOpenChange={() => setIsOpen(!isOpen)}>
            <Button onClick={() => setIsOpen(true)} className={cn(!updating && "w-full")}>{updating ? "Edit" : "New"}</Button>
            <DialogContent>
                <ScrollArea className="h-[34rem]">
                    <DialogHeader>
                        <DialogTitle>Create new item</DialogTitle>
                        <DialogDescription>
                            Create your new item
                        </DialogDescription>
                    </DialogHeader>
                    <Form {...form}>
                        <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8 mx-4">
                            <FormField
                                control={form.control}
                                name="name"
                                render={({ field }) => (
                                    <FormItem>
                                        <FormLabel>Name</FormLabel>
                                        <FormControl>
                                            <Input placeholder="Moose Horns" {...field} />
                                        </FormControl>
                                        <FormDescription>
                                            New item's name
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="description"
                                render={({ field }) => (
                                    <FormItem>
                                        <FormLabel>Description</FormLabel>
                                        <FormControl>
                                            <Input placeholder="These are the strongest best and coolest horns" {...field} />
                                        </FormControl>
                                        <FormDescription>
                                            New item's description
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="supplier"
                                render={({ field }) => (
                                    <FormItem className="flex flex-col">
                                        <FormLabel>Supplier</FormLabel>
                                        <Popover>
                                            <PopoverTrigger asChild>
                                                <FormControl>
                                                    <Button
                                                        variant="outline"
                                                        role="combobox"
                                                        className={cn(
                                                            "w-[200px] justify-between",
                                                            !field.value && "text-muted-foreground"
                                                        )}
                                                    >
                                                        {field.value
                                                            ? suppliers.find(
                                                                (supplier) => supplier.id === field.value
                                                            )?.name
                                                            : "Select supplier"}
                                                        <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
                                                    </Button>
                                                </FormControl>
                                            </PopoverTrigger>
                                            <PopoverContent className="w-[200px] p-0">
                                                <Command>
                                                    <CommandInput placeholder="Search supplier..." />
                                                    <CommandEmpty>No supplier found.</CommandEmpty>
                                                    <CommandGroup>
                                                        {suppliers.map((supplier) => (
                                                            <CommandItem
                                                                value={supplier.name}
                                                                key={supplier.id}
                                                                onSelect={() => {
                                                                    form.setValue("supplier", supplier.id)
                                                                }}
                                                            >
                                                                <Check
                                                                    className={cn(
                                                                        "mr-2 h-4 w-4",
                                                                        supplier.id === field.value
                                                                            ? "opacity-100"
                                                                            : "opacity-0"
                                                                    )}
                                                                />
                                                                {supplier.name}
                                                            </CommandItem>
                                                        ))}
                                                    </CommandGroup>
                                                </Command>
                                            </PopoverContent>
                                        </Popover>
                                        <FormDescription>
                                            This is the category of the item.
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="category"
                                render={({ field }) => (
                                    <FormItem className="flex flex-col">
                                        <FormLabel>Category</FormLabel>
                                        <Popover>
                                            <PopoverTrigger asChild>
                                                <FormControl>
                                                    <Button
                                                        variant="outline"
                                                        role="combobox"
                                                        className={cn(
                                                            "w-[200px] justify-between",
                                                            !field.value && "text-muted-foreground"
                                                        )}
                                                    >
                                                        {field.value
                                                            ? categories.find(
                                                                (category) => category.id === field.value
                                                            )?.name
                                                            : "Select category"}
                                                        <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
                                                    </Button>
                                                </FormControl>
                                            </PopoverTrigger>
                                            <PopoverContent className="w-[200px] p-0">
                                                <Command>
                                                    <CommandInput placeholder="Search category..." />
                                                    <CommandEmpty>No category found.</CommandEmpty>
                                                    <CommandGroup>
                                                        {categories.map((category) => (
                                                            <CommandItem
                                                                value={category.name}
                                                                key={category.id}
                                                                onSelect={() => {
                                                                    form.setValue("category", category.id)
                                                                }}
                                                            >
                                                                <Check
                                                                    className={cn(
                                                                        "mr-2 h-4 w-4",
                                                                        category.id === field.value
                                                                            ? "opacity-100"
                                                                            : "opacity-0"
                                                                    )}
                                                                />
                                                                {category.name}
                                                            </CommandItem>
                                                        ))}
                                                    </CommandGroup>
                                                </Command>
                                            </PopoverContent>
                                        </Popover>
                                        <FormDescription>
                                            This is the category of the item.
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="productionDate"
                                render={({ field }) => (
                                    <FormItem className="flex flex-col">
                                        <FormLabel>Production date</FormLabel>
                                        <Popover>
                                            <PopoverTrigger asChild>
                                                <Button
                                                    variant={"outline"}
                                                    className={cn(
                                                        "w-[280px] justify-start text-left font-normal",
                                                        !field.value && "text-muted-foreground"
                                                    )}
                                                >
                                                    <CalendarIcon className="mr-2 h-4 w-4" />
                                                    {field.value ? format(field.value, "PPP") : <span>Pick a date</span>}
                                                </Button>
                                            </PopoverTrigger>
                                            <PopoverContent className="w-auto p-0">
                                                <Calendar
                                                    mode="single"
                                                    selected={field.value}
                                                    onSelect={field.onChange}
                                                    initialFocus
                                                />
                                            </PopoverContent>
                                        </Popover>
                                        <FormDescription>
                                            Item's production date
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="expirationDate"
                                render={({ field }) => (
                                    <FormItem className="flex flex-col">
                                        <FormLabel>Expiration date</FormLabel>
                                        <Popover>
                                            <PopoverTrigger asChild>
                                                <Button
                                                    variant={"outline"}
                                                    className={cn(
                                                        "w-[280px] justify-start text-left font-normal",
                                                        !field.value && "text-muted-foreground"
                                                    )}
                                                >
                                                    <CalendarIcon className="mr-2 h-4 w-4" />
                                                    {field.value ? format(field.value, "PPP") : <span>Pick a date</span>}
                                                </Button>
                                            </PopoverTrigger>
                                            <PopoverContent className="w-auto p-0">
                                                <Calendar
                                                    mode="single"
                                                    selected={field.value}
                                                    onSelect={field.onChange}
                                                    initialFocus
                                                />
                                            </PopoverContent>
                                        </Popover>
                                        <FormDescription>
                                            Item's expiration date
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="storageCondition"
                                render={({ field }) => (
                                    <FormItem>
                                        <FormLabel>Storage condition</FormLabel>
                                        <FormControl>
                                            <Input placeholder="Store in a warm, dry place" {...field} />
                                        </FormControl>
                                        <FormDescription>
                                            Your item Storage condition
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="weight"
                                render={({ field }) => (
                                    <FormItem>
                                        <FormLabel>Weight</FormLabel>
                                        <FormControl>
                                            <Input placeholder="65"  {...field} />
                                        </FormControl>
                                        <FormDescription>
                                            New item's weight
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <FormField
                                control={form.control}
                                name="price"
                                render={({ field }) => (
                                    <FormItem>
                                        <FormLabel>Price</FormLabel>
                                        <FormControl>
                                            <Input placeholder="73" {...field} />
                                        </FormControl>
                                        <FormDescription>
                                            New item's price
                                        </FormDescription>
                                        <FormMessage />
                                    </FormItem>
                                )}
                            />
                            <div className="flex justify-between">
                                <Button type="button" variant="secondary">Cancel</Button>
                                <Button type="submit" >{updating ? "Save" : "Create"}</Button>
                            </div>
                        </form>
                    </Form>
                </ScrollArea>
            </DialogContent>
        </Dialog>

    );
}

export default ItemForm;