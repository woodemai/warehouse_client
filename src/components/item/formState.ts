export enum FormState {
    CREATE,
    UPDATE
}