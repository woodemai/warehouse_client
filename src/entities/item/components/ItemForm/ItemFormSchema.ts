import { ICategory } from "@/entities/category";
import { ISupplier } from "@/entities/supplier";
import * as z from "zod"

const supplierSchema =  z.object<ISupplier>({
    id: "",
    name: "",
    inn: 0,
});

const categorySchema = z.object<ICategory>({
    id: "",
    name: "",
    description: "",
});

const formSchema = z.object({
    name: z.string().min(2, {
        message: "Name must be at least 2 characters"
    }).max(30, {
        message: "Name must be less than 30 characters"
    }),
    description: z.string().min(6, {
        message: "Description must be at least 6 characters"
    }).max(300, {
        message: "Description must be less than 300 characters"
    }),
    productionDate: z.date().min(new Date("2020-01-01"), { message: "Too old" }),
    expirationDate: z.date().min(new Date("2020-01-01"), { message: "Too old" }),
    storageCondition: z.string().min(8, {
        message: "Storage condition must be at least 8 characters"
    }).max(100, {
        message: "Storage condition must be less than 100 characters"
    }),
    weight: z.coerce.number(),
    price: z.coerce.number(),
    supplier: z.string(),
    category: z.string()
})
export default formSchema;