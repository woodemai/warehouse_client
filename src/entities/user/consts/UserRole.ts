export enum UserRole {
    EMPLOYEE = "EMPLOYEE",
    SUPPLIER = "SUPPLIER",
    BUYER = "BUYER"
}