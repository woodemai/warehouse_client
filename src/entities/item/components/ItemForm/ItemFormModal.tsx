import { Button } from "@/shared/components/ui/button"
import { DialogContent, DialogDescription, DialogHeader, DialogTitle } from "@/shared/components/ui/dialog"
import { Form, FormControl, FormDescription, FormField, FormItem, FormLabel, FormMessage } from "@/shared/components/ui/form"
import { Input } from "@/shared/components/ui/input"
import { ScrollArea } from "@/shared/components/ui/scroll-area"
import { FormState } from "@/shared/consts/formState"
import { FC, useEffect, useState } from "react"
import { z } from "zod"
import formSchema from "./ItemFormSchema"
import { UseFormReturn } from "react-hook-form"
import { Popover, PopoverContent, PopoverTrigger } from "@/shared/components/ui/popover"
import { Calendar } from "@/shared/components/ui/calendar"
import { CalendarIcon, Check, ChevronsUpDown } from "lucide-react"
import { cn } from "@/shared/lib/shadcn/utils"
import { Command, CommandEmpty, CommandGroup, CommandInput, CommandItem } from "@/shared/components/ui/command"
import { Textarea } from "@/shared/components/ui/textarea"
import { CategoryService, ICategory } from "@/entities/category"
import { ISupplier, SupplierService } from "@/entities/supplier"
import { format } from "date-fns"

interface ItemFormModalProps {
    formState: FormState,
    onSubmit: (values: z.infer<typeof formSchema>) => void,
    setIsOpen: (open: boolean) => void,
    form: UseFormReturn<{
        name: string;
        description: string;
        supplier: {
            id: string;
            name: string;
            inn: number;
        };
        productionDate: Date;
        expirationDate: Date;
        storageCondition: string;
        weight: number;
        price: number;
        category: {
            id: string,
            name: string,
            description: string
        };
    }, undefined>
}

const ItemFormModal: FC<ItemFormModalProps> = ({
    formState,
    onSubmit,
    form,
    setIsOpen
}) => {
    const [categories, setCategories] = useState<ICategory[]>([]);
    const [suppliers, setSuppliers] = useState<ISupplier[]>([]);

    useEffect(() => {
        CategoryService.getCategorires().then(res => setCategories(res.data));
        SupplierService.getSuppliers().then(res => setSuppliers(res.data));
    }, []);
    return (
        <DialogContent className="bg-white">
            <ScrollArea className="h-[34rem]">
                <DialogHeader>
                    <DialogTitle>Создать новый предмет</DialogTitle>
                    <DialogDescription>
                        Введите характеристики ниже
                    </DialogDescription>
                </DialogHeader>
                <Form {...form}>
                    <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-8 mx-4">
                        <FormField
                            control={form.control}
                            name="name"
                            render={({ field }) => (
                                <FormItem>
                                    <FormLabel>Название</FormLabel>
                                    <FormControl>
                                        <Input autoComplete="name" placeholder="Копыта" {...field} />
                                    </FormControl>
                                    <FormDescription>
                                        Название нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="description"
                            render={({ field }) => (
                                <FormItem>
                                    <FormLabel>Описание</FormLabel>
                                    <FormControl>
                                        <Textarea placeholder="Самые лошадиные копыта на диком западе" {...field} />
                                    </FormControl>
                                    <FormDescription>
                                        Описание нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="supplier"
                            render={({ field }) => (
                                <FormItem className="flex flex-col">
                                    <FormLabel>Поставщик</FormLabel>
                                    <Popover>
                                        <PopoverTrigger asChild>
                                            <FormControl>
                                                <Button
                                                    variant="outline"
                                                    role="combobox"
                                                    className={cn(
                                                        "w-auto justify-between",
                                                        !field.value && "text-muted-foreground"
                                                    )}
                                                >
                                                    {field.value
                                                        ? suppliers.find(
                                                            (supplier) => supplier === field.value
                                                        )?.name
                                                        : "Выберите поставщика"}
                                                    <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
                                                </Button>
                                            </FormControl>
                                        </PopoverTrigger>
                                        <PopoverContent className="w-[200px] p-0">
                                            <Command>
                                                <CommandInput placeholder="Найти поставщика..." />
                                                <CommandEmpty>Поставщик не найден</CommandEmpty>
                                                <CommandGroup>
                                                    {suppliers.map((supplier) => (
                                                        <CommandItem
                                                            value={supplier.name}
                                                            key={supplier.id}
                                                            onSelect={() => {
                                                                form.setValue("supplier", supplier)
                                                            }}
                                                        >
                                                            <Check
                                                                className={cn(
                                                                    "mr-2 h-4 w-4",
                                                                    supplier === field.value
                                                                        ? "opacity-100"
                                                                        : "opacity-0"
                                                                )}
                                                            />
                                                            {supplier.name}
                                                        </CommandItem>
                                                    ))}
                                                </CommandGroup>
                                            </Command>
                                        </PopoverContent>
                                    </Popover>
                                    <FormDescription>
                                        Поставщик нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="category"
                            render={({ field }) => (
                                <FormItem className="flex flex-col">
                                    <FormLabel>Категория</FormLabel>
                                    <Popover>
                                        <PopoverTrigger asChild>
                                            <FormControl>
                                                <Button
                                                    variant="outline"
                                                    role="combobox"
                                                    className={cn(
                                                        "w-auto justify-between",
                                                        !field.value && "text-muted-foreground"
                                                    )}
                                                >
                                                    {field.value
                                                        ? categories.find(
                                                            (category) => category === field.value
                                                        )?.name
                                                        : "Выберите категорию"}
                                                    <ChevronsUpDown className="ml-2 h-4 w-4 shrink-0 opacity-50" />
                                                </Button>
                                            </FormControl>
                                        </PopoverTrigger>
                                        <PopoverContent className="w-[200px] p-0">
                                            <Command>
                                                <CommandInput placeholder="Найти категорию..." />
                                                <CommandEmpty>Категория не найдена.</CommandEmpty>
                                                <CommandGroup>
                                                    {categories.map((category) => (
                                                        <CommandItem
                                                            value={category.name}
                                                            key={category.id}
                                                            onSelect={() => {
                                                                form.setValue("category", category)
                                                            }}
                                                        >
                                                            <Check
                                                                className={cn(
                                                                    "mr-2 h-4 w-4",
                                                                    category === field.value
                                                                        ? "opacity-100"
                                                                        : "opacity-0"
                                                                )}
                                                            />
                                                            {category.name}
                                                        </CommandItem>
                                                    ))}
                                                </CommandGroup>
                                            </Command>
                                        </PopoverContent>
                                    </Popover>
                                    <FormDescription>
                                        Категория нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="productionDate"
                            render={({ field }) => (
                                <FormItem className="flex flex-col">
                                    <FormLabel htmlFor={field.name}>Дата производства</FormLabel>
                                    <Popover>
                                        <PopoverTrigger asChild>
                                            <Button
                                                id={field.name}
                                                variant={"outline"}
                                                className={cn(
                                                    "w-[280px] justify-start text-left font-normal",
                                                    !field.value && "text-muted-foreground"
                                                )}
                                            >
                                                <CalendarIcon className="mr-2 h-4 w-4" />
                                                {field.value ? format(field.value, "PPP") : <span>Выберите дату</span>}
                                            </Button>
                                        </PopoverTrigger>
                                        <PopoverContent className="w-auto p-0">
                                            <Calendar
                                                mode="single"
                                                selected={field.value}
                                                onSelect={field.onChange}
                                                initialFocus
                                            />
                                        </PopoverContent>
                                    </Popover>
                                    <FormDescription>
                                        Дата производства нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="expirationDate"
                            render={({ field }) => (
                                <FormItem className="flex flex-col">
                                    <FormLabel htmlFor={field.name}>Срок годности</FormLabel>
                                    <Popover>
                                        <PopoverTrigger asChild>
                                            <Button
                                                id={field.name}
                                                variant={"outline"}
                                                className={cn(
                                                    "w-[280px] justify-start text-left font-normal",
                                                    !field.value && "text-muted-foreground"
                                                )}
                                            >
                                                <CalendarIcon className="mr-2 h-4 w-4" />
                                                {field.value ? format(field.value, "PPP") : <span>Выберите дату</span>}
                                            </Button>
                                        </PopoverTrigger>
                                        <PopoverContent className="w-auto p-0">
                                            <Calendar
                                                mode="single"
                                                selected={field.value}
                                                onSelect={field.onChange}
                                                initialFocus
                                            />
                                        </PopoverContent>
                                    </Popover>
                                    <FormDescription>
                                        Годен до
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="storageCondition"
                            render={({ field }) => (
                                <FormItem>
                                    <FormLabel>Условия хранения</FormLabel>
                                    <FormControl>
                                        <Input placeholder="Хранить в теплом сухом месте" {...field} />
                                    </FormControl>
                                    <FormDescription>
                                        Условия хранения нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="weight"
                            render={({ field }) => (
                                <FormItem>
                                    <FormLabel>Вес</FormLabel>
                                    <FormControl>
                                        <Input placeholder="65кг"  {...field} />
                                    </FormControl>
                                    <FormDescription>
                                        Вес нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <FormField
                            control={form.control}
                            name="price"
                            render={({ field }) => (
                                <FormItem>
                                    <FormLabel>Цена</FormLabel>
                                    <FormControl>
                                        <Input placeholder="1999 ₽" {...field} />
                                    </FormControl>
                                    <FormDescription>
                                        Цена нового предмета
                                    </FormDescription>
                                    <FormMessage />
                                </FormItem>
                            )}
                        />
                        <div className="flex justify-between">
                            <Button type="button" variant="secondary" onClick={() => setIsOpen(false)}>Отмена</Button>
                            <Button type="submit" >{formState === FormState.UPDATE ? "Обновить" : "Создать"}</Button>
                        </div>
                    </form>
                </Form>
            </ScrollArea>
        </DialogContent>
    )
}
export default ItemFormModal;