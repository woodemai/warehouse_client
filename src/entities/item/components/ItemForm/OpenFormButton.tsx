import { Button } from "@/shared/components/ui/button"
import { FormState } from "@/shared/consts/formState"
import { cn } from "@/shared/lib/shadcn/utils"
import { FC } from "react"

interface OpenFormButtonProps {
    formState: FormState
    setIsOpen: (isOpen: boolean) => void
}
const OpenFormButton: FC<OpenFormButtonProps> = ({
    formState,
    setIsOpen
}) => {
    return (
        <Button data-testid="open-form-button" onClick={() => setIsOpen(true)}
            className={cn(formState === FormState.CREATE && "w-full")}>
            {formState === FormState.UPDATE ? "Изменить" : "Создать"}
        </Button>
    )
}
export default OpenFormButton;