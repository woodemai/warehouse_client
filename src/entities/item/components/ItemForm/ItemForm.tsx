import { Dialog } from "@/shared/components/ui/dialog"
import { useForm } from "react-hook-form"
import { zodResolver } from "@hookform/resolvers/zod"
import * as z from "zod"
import { FC, Suspense, lazy, useCallback, useState } from "react"
import formSchema from "./ItemFormSchema";
import { IItem } from "@/entities/item/models/IItem"
import { FormState } from "../../../../shared/consts/formState"
import { ItemService } from "../../api/ItemService"
import OpenFormButton from "./OpenFormButton"
import { Loader } from "@/shared/components/ui/loader"
const ItemFormModal = lazy(() => import("./ItemFormModal"));

interface ItemFormProps {
    item?: IItem,
    formState: FormState
    setUpdated?: (updated: boolean) => void
}
export const ItemForm: FC<ItemFormProps> = ({
    item,
    formState,
    setUpdated
}) => {

    const [isOpen, setIsOpen] = useState(false);
    const form = useForm<z.infer<typeof formSchema>>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            name: item?.name ?? "",
            description: item?.description ?? "",
            storageCondition: item?.storageCondition ?? "",
            productionDate: new Date(),
            expirationDate: new Date(),
            weight: item?.weight ?? 0,
            price: item?.price ?? 0,
            supplier: item?.supplier.id ?? "",
            category: item?.category.id ?? "",
        },
    })
    const onSubmit = useCallback((values: z.infer<typeof formSchema>) => {
        if (formState === FormState.UPDATE && item) {
            ItemService.updateItem(values, item.id)
                .then(() => {
                    form.reset()
                    setIsOpen(false)
                })
        } else if (setUpdated) {
            ItemService.createItem(values)
                .then(() => {
                    form.reset()
                    setIsOpen(false)
                })
                .finally(() => setUpdated(true))
        }
    }, [form, formState, item, setUpdated])
    return (
        <Dialog open={isOpen} onOpenChange={() => setIsOpen(!isOpen)}>
            <OpenFormButton setIsOpen={setIsOpen} formState={formState} />
            <Suspense fallback={<Loader />}>
                <ItemFormModal setIsOpen={setIsOpen} form={form} onSubmit={onSubmit} formState={formState} />
            </Suspense>
        </Dialog>

    );
}