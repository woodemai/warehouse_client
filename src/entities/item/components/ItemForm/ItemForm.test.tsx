import { test, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { ItemForm } from './ItemForm'
import { FormState } from '@/shared/consts/formState'
import "@testing-library/jest-dom";


test("button should be in the document", () => {
    const { container } = render(<ItemForm formState={FormState.CREATE} />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId(/open-form-button/i)).toBeInTheDocument();
});