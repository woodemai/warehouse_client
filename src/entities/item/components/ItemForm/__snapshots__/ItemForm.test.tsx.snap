// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`button should be in the document 1`] = `
<div>
  <button
    class="inline-flex items-center justify-center rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full"
    data-testid="open-form-button"
  >
    Создать
  </button>
  <div
    class="text-3xl w-full h-screen flex justify-center items-center"
  >
    Loading...
  </div>
</div>
`;
